File: app/channels/chat_channel.rb
class ChatChannel < ApplicationCable::Channel
  def subscribed
    chat_room_id = params[:chat_room_id]
    stream_from "chat_#{chat_room_id}"
  end

  def unsubscribed
    # Opcional: lógica de limpeza quando o usuário sai do canal
    # p "Usuário #{current_user.email} desconectado do chat #{params[:chat_room_id]}"
  end

  def receive(data)
    chat_room_id = params[:chat_room_id]

    # p "dentro do chatChannel #{current_user.id}"
    message = Message.create!(
      user: current_user,
      chat_room_id: chat_room_id,
      content: data["content"]
    )

    ActionCable.server.broadcast(
      "chat_#{chat_room_id}",
      {
        content: message.content,
        user_email: message.user.email,
        user_id: message.user.id
      }
    )
  end

  private

  def render_message(message)
    ApplicationController.renderer.render(partial: "messages/message", locals: { message: message })
  end
end

File: app/javascript/channels/chat_channel.js
import consumer from "./consumer"

document.addEventListener('turbolinks:load', () => {
  const chatContainer = document.getElementById('chat-container')
  
  if (chatContainer) {
    const chatRoomId = chatContainer.dataset.chatroomId
    const currentUserId = chatContainer.dataset.currentUserId
    // console.log(`current_user no chat_channel.js ${currentUserId}`);

    consumer.subscriptions.create(
      {
        channel: "ChatChannel",
        chat_room_id: chatRoomId
      },
      {
        received(data) {
          const messageContainer = document.getElementById('messages')
          const isCurrentUser = data.user_id.toString() === currentUserId
          
          // Criar o HTML da mensagem com base em quem enviou
          const messageHTML = `
            <div class="message mb-2 ${isCurrentUser ? 'text-end' : 'text-start'}">
              <div class="d-inline-block p-2 rounded ${isCurrentUser ? 'bg-primary text-white' : 'bg-light'}" style="max-width: 70%;">
                <small class="d-block ${isCurrentUser ? 'text-white-50' : 'text-muted'}">
                  ${data.user_email}
                </small>
                ${data.content}
              </div>
            </div>
          `
          
          messageContainer.insertAdjacentHTML('beforeend', messageHTML)
          messageContainer.scrollTop = messageContainer.scrollHeight
        }
      }
    )
  }

  document.addEventListener('turbolinks:load', () => {
    const sendButton = document.getElementById('send-button')
    const chatInput = document.getElementById('chat-input')
    const chatContainer = document.getElementById('chat-container')
  
    if (sendButton && chatInput && chatContainer) {
      const chatRoomId = chatContainer.dataset.chatroomId
      const subscription = consumer.subscriptions.create(
        { channel: "ChatChannel", chat_room_id: chatRoomId },
        { received: data => {} }  // handled above
      )
  
      const sendMessage = () => {
        const content = chatInput.value.trim()
        if (content) {
          subscription.send({ content: content })
          chatInput.value = ''
        }
      }
  
      sendButton.addEventListener('click', sendMessage)
      
      chatInput.addEventListener('keypress', (e) => {
        if (e.key === 'Enter' && !e.shiftKey) {
          e.preventDefault()
          sendMessage()
        }
      })
    }
  })
})


